lib/segment/src/common/rocksdb_operations.rs

```diff
@@ -12,70 +15,274 @@ pub const DB_PAYLOAD_CF: &str = "payload";
+                OperationError::service_error(&format!("RocksDB put_cf error: {}", err))
+            })?;
+        if self.put_fixed_key.load(Ordering::Relaxed) {
+            db.db
+                .put_cf_opt(
+                    cf_handle,
+                    FIXED_KEY,
+                    FIXED_VALUE,
+                    &Self::get_write_options(),
+                )
+                .map_err(|err| {
+                    OperationError::service_error(&format!(
+                        "RocksDB put_cf (fixed key) error: {}",
+                        err
+                    ))
+                })?;
+            self.put_fixed_key.store(false, Ordering::Relaxed);
+        }
+        Ok(())
+    }
+
+    pub fn get_pinned<T, F>(&self, key: &[u8], f: F) -> OperationResult<Option<T>>
+    where
+        F: FnOnce(&[u8]) -> T,
+    {
+        let db = self.database.borrow();
+        let cf_handle = self.get_column_family(&db)?;
+        let result = db
+            .db
+            .get_pinned_cf(cf_handle, key)
+            .map_err(|err| {
+                OperationError::service_error(&format!("RocksDB get_pinned_cf error: {}", err))
+            })?
+            .map(|value| f(&value));
+        Ok(result)
+    }
+
+    pub fn remove<K>(&self, key: K) -> OperationResult<()>
+    where
+        K: AsRef<[u8]>,
+    {
+        let db = self.database.borrow();
+        let cf_handle = self.get_column_family(&db)?;
+        db.db.delete_cf(cf_handle, key).map_err(|err| {
+            OperationError::service_error(&format!("RocksDB delete_cf error: {}", err))
+        })?;
+        Ok(())
+    }
+
+    pub fn iter<'a: 'b, 'b>(
+        &self,
+    ) -> OperationResult<ArcAtomicRefCellIterator<Database, DatabaseColumnIterator<'a, 'b>>> {
+        let db = self.database.borrow();
+        // check column existing
+        self.get_column_family(&db)?;
+        let column_name = self.column_name.clone();
+        Ok(ArcAtomicRefCellIterator::new(
+            self.database.clone(),
+            move |db| {
+                // column existsing is already checked
+                let handle = db.db.cf_handle(&column_name).unwrap();
+                let mut iter = db.db.raw_iterator_cf(&handle);
+                iter.seek_to_first();
+                DatabaseColumnIterator {
+                    handle,
+                    iter,
+                    just_seeked: true,
+                }
+            },
+        ))
+    }
+
+    pub fn flush(&self) -> OperationResult<()> {
+        let db = self.database.borrow();
+        let column_family = self.get_column_family(&db)?;
 
-    if db_mut.cf_handle(store_cf_name).is_some() {
-        db_mut.drop_cf(store_cf_name)?;
+        db.db.flush_cf(column_family).map_err(|err| {
+            OperationError::service_error(&format!("RocksDB flush_cf error: {}", err))
+        })?;
+        if db.allow_add_new_keys {
+            self.put_fixed_key.store(true, Ordering::Relaxed);
+        }
```

Comment:
Looks like an overkill to do it on each flush. Let's maybe do in on each n-th (n=64?) flush.

Reply:
In case of 1 write between flushes we'll get sst file for each write. That's why I set this flag each flush. And flush takes at least 50 milliseconds, so I think the impact of setting this flag is too low

Reply:
I am more concert of the compaction impact

Reply:
I tested key writing every 10 flushes. Compaction works fine, files are mostly compacted. But there were old small non-compacted files that were still uncompacted. I propose to keep as-is because with such solution all small files are compacted. And when we get into trouble with compaction performance, we could go back to key-value writing each 64? flushes.